Electra spec changes for `v1.5.0-beta.0`

michaelsproul opened the pull request:
## Proposed Changes

Consensus changes for `v1.5.0-alpha.10`, and consequently `v1.5.0-beta.0` which is only cosmetically different.

Built on:

- https://github.com/sigp/lighthouse/pull/6603
- https://github.com/sigp/lighthouse/pull/6329

pawanjay176 committed:
First pass

pawanjay176 committed:
Add restrictions to RuntimeVariableList api

pawanjay176 committed:
Use empty_uninitialized and fix warnings

pawanjay176 committed:
Fix some todos

pawanjay176 committed:
Merge branch 'unstable' into max-blobs-preset

pawanjay176 committed:
Fix take impl on RuntimeFixedList

pawanjay176 committed:
cleanup

pawanjay176 committed:
Fix test compilations

pawanjay176 committed:
Fix some more tests

pawanjay176 committed:
Fix test from unstable

pawanjay176 committed:
Merge branch 'unstable' into max-blobs-preset

michaelsproul committed:
Implement "Bugfix and more withdrawal tests"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=6731) <br/>All committers have signed the CLA.

michaelsproul committed:
Implement "Add missed exit checks to consolidation processing"

michaelsproul committed:
Implement "Update initial earliest_exit_epoch calculation"

michaelsproul committed:
Implement "Limit consolidating balance by validator.effective_balance"

michaelsproul committed:
Implement "Use 16-bit random value in validator filter"

michaelsproul committed:
Implement "Do not change creds type on consolidation"

michaelsproul commented:
I think I've made all the substantial changes for Electra, but the tests are broken and I haven't looked into why. Some of them probably need disabling because this branch doesn't have the PeerDAS changes.

I'll be back on Jan 6.

eserilev committed:
Rename PendingPartialWithdraw index field to validator_index

eserilev committed:
Skip slots to get test to pass and add TODO

eserilev committed:
Implement "Synchronously check all transactions to have non-zero length"

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into max-blobs-preset

michaelsproul committed:
Remove footgun function

michaelsproul committed:
Minor simplifications

michaelsproul committed:
Move from preset to config

michaelsproul committed:
Fix typo

michaelsproul committed:
Revert "Remove footgun function"

This reverts commit de01f923c7452355c87f50c0e8031ca94fa00d36.

michaelsproul committed:
Try fixing tests

eserilev committed:
Implement "bump minimal preset MAX_BLOB_COMMITMENTS_PER_BLOCK and KZG_COMMITMENT_INCLUSION_PROOF_DEPTH"

michaelsproul committed:
Thread through ChainSpec

michaelsproul committed:
Fix release tests

michaelsproul committed:
Move RuntimeFixedVector into module and rename

michaelsproul committed:
Add test

eserilev committed:
Implement "Remove post-altair `initialize_beacon_state_from_eth1` from specs"

michaelsproul committed:
Update preset YAML

pawanjay176 committed:
Remove empty RuntimeVarList awefullness

michaelsproul committed:
Make max_blobs_per_block a config parameter (#6329)

Squashed commit of the following:

commit 04b3743ec1e0b650269dd8e58b540c02430d1c0d
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 17:36:58 2025 +1100

    Add test

commit 440e85419940d4daba406d910e7908dd1fe78668
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 17:24:50 2025 +1100

    Move RuntimeFixedVector into module and rename

commit f66e179a40c3917eee39a93534ecf75480172699
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 17:17:17 2025 +1100

    Fix release tests

commit e4bfe71cd1f0a2784d0bd57f85b2f5d8cf503ac1
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 17:05:30 2025 +1100

    Thread through ChainSpec

commit 063b79c16abd3f6df47b85efcf3858177bc933b9
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 15:32:16 2025 +1100

    Try fixing tests

commit 88bedf09bc647de66bd1ff944bbc8fb13e2b7590
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 15:04:37 2025 +1100

    Revert "Remove footgun function"

    This reverts commit de01f923c7452355c87f50c0e8031ca94fa00d36.

commit 32483d385b66f252d50cee5b524e2924157bdcd4
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 15:04:32 2025 +1100

    Fix typo

commit 2e86585b478c012f6e3483989c87e38161227674
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 15:04:15 2025 +1100

    Move from preset to config

commit 1095d60a40be20dd3c229b759fc3c228b51e51e3
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 14:38:40 2025 +1100

    Minor simplifications

commit de01f923c7452355c87f50c0e8031ca94fa00d36
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 14:06:57 2025 +1100

    Remove footgun function

commit 0c2c8c42245c25b8cf17885faf20acd3b81140ec
Merge: 21ecb58ff f51a292f7
Author: Michael Sproul <michael@sigmaprime.io>
Date:   Mon Jan 6 14:02:50 2025 +1100

    Merge remote-tracking branch 'origin/unstable' into max-blobs-preset

commit f51a292f77575a1786af34271fb44954f141c377
Author: Daniel Knopik <107140945+dknopik@users.noreply.github.com>
Date:   Fri Jan 3 20:27:21 2025 +0100

    fully lint only explicitly to avoid unnecessary rebuilds (#6753)

    * fully lint only explicitly to avoid unnecessary rebuilds

commit 7e0cddef321c2a069582c65b58e5f46590d60c49
Author: Akihito Nakano <sora.akatsuki@gmail.com>
Date:   Tue Dec 24 10:38:56 2024 +0900

    Make sure we have fanout peers when publish (#6738)

    * Ensure that `fanout_peers` is always non-empty if it's `Some`

commit 21ecb58ff88b86435ab62d9ac227394c10fdcd22
Merge: 2fcb2935e 9aefb5539
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Mon Oct 21 14:46:00 2024 -0700

    Merge branch 'unstable' into max-blobs-preset

commit 2fcb2935ec7ef4cd18bbdd8aedb7de61fac69e61
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Fri Sep 6 18:28:31 2024 -0700

    Fix test from unstable

commit 12c6ef118a1a6d910c48d9d4b23004f3609264c7
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Wed Sep 4 16:16:36 2024 -0700

    Fix some more tests

commit d37733b846ce58e318e976d6503ca394b4901141
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Wed Sep 4 12:47:36 2024 -0700

    Fix test compilations

commit 52bb581e071d5f474d519366e860a4b3a0b52f78
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Tue Sep 3 18:38:19 2024 -0700

    cleanup

commit e71020e3e613910e0315f558ead661b490a0ff20
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Tue Sep 3 17:16:10 2024 -0700

    Fix take impl on RuntimeFixedList

commit 13f9bba6470b2140e5c34f14aed06dab2b062c1c
Merge: 60100fc6b 4e675cf5d
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Tue Sep 3 16:08:59 2024 -0700

    Merge branch 'unstable' into max-blobs-preset

commit 60100fc6be72792ff33913d7e5a53434c792aacf
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Fri Aug 30 16:04:11 2024 -0700

    Fix some todos

commit a9cb329a221a809f7dd818984753826f91c2e26b
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Fri Aug 30 15:54:00 2024 -0700

    Use empty_uninitialized and fix warnings

commit 4dc6e6515ecf75cefa4de840edc7b57e76a8fc9e
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Fri Aug 30 15:53:18 2024 -0700

    Add restrictions to RuntimeVariableList api

commit 25feedfde348b530c4fa2348cc71a06b746898ed
Author: Pawan Dhananjay <pawandhananjay@gmail.com>
Date:   Thu Aug 29 16:11:19 2024 -0700

    First pass

pawanjay176 committed:
Fix tests

michaelsproul committed:
Implement max_blobs_per_block_electra

michaelsproul commented on consensus/fork_choice/tests/tests.rs:
```diff
@@ -1226,6 +1226,10 @@ async fn progressive_balances_cache_attester_slashing() {
+        // TODO(electra) The shuffling calculations changed between Altair and Electra. Without
+        // skipping slots this test breaks. For some reason `fork_name_unchecked` returns Altair
+        // initially, even though this test harness should be initialized with the most recent fork, i.e. Electra
+        .skip_slots(32)
```
TODO: I should look into this

michaelsproul committed:
Fix config issues

michaelsproul committed:
Simplify BlobSidecarListFromRoot

michaelsproul committed:
Disable PeerDAS tests

michaelsproul commented:
I've squash merged the max blobs PR:

- https://github.com/sigp/lighthouse/pull/6329

We should merge that to `unstable` first.

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into max-blobs-preset

michaelsproul committed:
Bump quota to account for new target (6)

michaelsproul committed:
Remove clone

pawanjay176 committed:
Fix issue from review

pawanjay176 committed:
Try to remove ugliness

pawanjay176 committed:
Merge branch 'unstable' into max-blobs-preset

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into electra-alpha10

michaelsproul committed:
Merge commit '04b3743ec1e0b650269dd8e58b540c02430d1c0d' into electra-alpha10

michaelsproul committed:
Merge remote-tracking branch 'pawan/max-blobs-preset' into electra-alpha10

michaelsproul committed:
Update tests to v1.5.0-beta.0

eserilev committed:
Resolve merge conflicts

eserilev committed:
Linting

eserilev committed:
fmt

eserilev committed:
Fix test and add TODO

eserilev commented on consensus/fork_choice/tests/tests.rs:
```diff
@@ -1261,6 +1261,10 @@ async fn progressive_balances_cache_proposer_slashing() {
         .apply_blocks_while(|_, state| state.finalized_checkpoint().epoch == 0)
         .await
         .unwrap()
+        // TODO(electra) The shuffling calculations changed between Altair and Electra. Without
+        // skipping slots this test breaks. For some reason `fork_name_unchecked` returns Altair
+        // initially, even though this test harness should be initialized with the most recent fork, i.e. Electra
+        .skip_slots(32)
```
another TODO we should take a look at

michaelsproul replied:
I've fixed both these TODOs by making the test harness skip block proposals for slashed validators. Commit: https://github.com/sigp/lighthouse/pull/6731/commits/f4fe1b8475365f1f15ba5f2878896c903c18d620

michaelsproul committed:
Gracefully handle slashed proposers in fork choice tests

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into electra-alpha10

michaelsproul commented:
This is ready for review once CI passes (I think it should :crossed_fingers: )

michaelsproul commented on beacon_node/lighthouse_network/src/rpc/codec.rs:
```diff
@@ -590,16 +590,14 @@ fn handle_rpc_request<E: EthSpec>(
             // TODO(pawan): change this to max_blobs_per_rpc_request in the alpha10 PR
-            if max_requested_blobs > spec.max_request_blob_sidecars {
+            let max_blobs = spec.max_blobs_per_block_by_fork(current_fork);
+            if req.count > max_blobs {
```
This got merged incorrectly, I'll fix it

michaelsproul replied:
Fixed in https://github.com/sigp/lighthouse/pull/6731/commits/3b177327cf2a31b738688f863b812ddf0405f149, and then other merge conflict regressions fixed in https://github.com/sigp/lighthouse/pull/6731/commits/e2ff44047125228d76e1ef210f453b4711620f27.

michaelsproul committed:
Keep latest changes from max_blobs_per_block PR in codec.rs

michaelsproul commented on consensus/types/presets/mainnet/phase0.yaml:
```diff
@@ -85,4 +85,4 @@ MAX_ATTESTATIONS: 128
 MAX_DEPOSITS: 16
 # 2**4 (= 16)
-MAX_VOLUNTARY_EXITS: 16
+MAX_VOLUNTARY_EXITS: 16
```
Note for reviewers: the contents of the presets and configs are copied verbatim from `consensus-specs`, which is why there are a mix of substantive changes and formatting changes.

michaelsproul committed:
Revert a few more regressions and add a comment

michaelsproul reviewed (commented):
Self-review complete. I think this is good now.

michaelsproul commented:
Eh, tests are broken. Working on it.

michaelsproul committed:
Disable more DAS tests

michaelsproul committed:
Improve validator monitor test a little

michaelsproul committed:
Make test more robust

michaelsproul committed:
Fix sync test that didn't understand blobs

michaelsproul commented:
Should be good now.

dapplion commented on consensus/types/src/beacon_state.rs:
```diff
@@ -953,6 +972,21 @@ impl<E: EthSpec> BeaconState<E> {
 
+    /// Get two random bytes from the given `seed`.
+    ///
+    /// This is used in place of the
```
cropped comment

dapplion reviewed (approved):
These changes are in parity with the spec 👍

michaelsproul committed:
Fill out cropped comment

michaelsproul commented:
Thanks for the reviews guys! I'll merge this so we can get closer to devnet 5 support in `unstable`.

michaelsproul commented:
@mergify queue
